include/nbl/builtin/hlsl/math/morton.hlsl

```diff
@@ -0,0 +1,283 @@
+inline T morton2d_decode(T x)
+{
+    x = x & morton2d_mask<T>(0);
+    x = (x | (x >> 1)) & morton2d_mask<T>(1);
+    x = (x | (x >> 2)) & morton2d_mask<T>(2);
+    if constexpr (bitDepth > 8u)
+    {
+        x = (x | (x >> 4)) & morton2d_mask<T>(3);
+    }
+    if constexpr (bitDepth > 16u)
+    {
+        x = (x | (x >> 8)) & morton2d_mask<T>(4);
+    }
+    if constexpr (bitDepth > 32u)
+    {
+        x = (x | (x >> 16));
+    }
+    return x;
+}
+
+//! Puts bits on even positions filling gaps with 0s
+template <typename T, uint32_t bitDepth>
+inline T separate_bits_2d(T x)
+{
+    if constexpr (bitDepth > 32u)
+    {
+        x = (x | (x << 16)) & morton2d_mask<T>(4);
+    }
+    if constexpr (bitDepth > 16u)
+    {
+        x = (x | (x << 8)) & morton2d_mask<T>(3);
+    }
+    if constexpr (bitDepth > 8u)
+    {
+        x = (x | (x << 4)) & morton2d_mask<T>(2);
+    }
+    x = (x | (x << 2)) & morton2d_mask<T>(1);
+    x = (x | (x << 1)) & morton2d_mask<T>(0);
+
+    return x;
+}
+template <typename T, uint32_t bitDepth>
+inline T separate_bits_3d(T x)
+{
+    if constexpr (bitDepth > 32u)
+    {
+        x = (x | (x << 32)) & morton3d_mask<T>(4);
+    }
+    if constexpr (bitDepth > 16u)
+    {
+        x = (x | (x << 16)) & morton3d_mask<T>(3);
+    }
+    if constexpr (bitDepth > 8u)
+    {
+        x = (x | (x << 8)) & morton3d_mask<T>(2);
+    }
+    x = (x | (x << 4)) & morton3d_mask<T>(1);
+    x = (x | (x << 2)) & morton3d_mask<T>(0);
+
+    return x;
+}
+template <typename T, uint32_t bitDepth>
+inline T separate_bits_4d(T x)
+{
+    if constexpr (bitDepth > 32u)
+    {
+        x = (x | (x << 24)) & morton4d_mask<T>(3);
+    }
+    if constexpr (bitDepth > 16u)
+    {
+        x = (x | (x << 12)) & morton4d_mask<T>(2);
+    }
+    if constexpr (bitDepth > 8u)
+    {
+        x = (x | (x << 6)) & morton4d_mask<T>(1);
+    }
+    x = (x | (x << 3)) & morton4d_mask<T>(0);
+
+    return x;
+}
+#endif
+}
+
+template<typename T, uint32_t bitDepth = sizeof(T) * 8u>
+T morton2d_decode_x(T _morton) { return impl::morton2d_decode<T, bitDepth>(_morton); }
+template<typename T, uint32_t bitDepth = sizeof(T) * 8u>
+T morton2d_decode_y(T _morton) { return impl::morton2d_decode<T, bitDepth>(_morton >> 1); }
+
+template<typename T, uint32_t bitDepth = sizeof(T) * 8u>
+T morton2d_encode(T x, T y) { return impl::separate_bits_2d<T, bitDepth>(x) | (impl::separate_bits_2d<T, bitDepth>(y) << 1); }
+template<typename T, uint32_t bitDepth = sizeof(T) * 8u>
+T morton3d_encode(T x, T y, T z) { return impl::separate_bits_3d<T, bitDepth>(x) | (impl::separate_bits_3d<T, bitDepth>(y) << 1) | (impl::separate_bits_3d<T, bitDepth>(z) << 2); }
+template<typename T, uint32_t bitDepth = sizeof(T) * 8u>
+T morton4d_encode(T x, T y, T z, T w) { return impl::separate_bits_4d<T, bitDepth>(x) | (impl::separate_bits_4d<T, bitDepth>(y) << 1) | (impl::separate_bits_4d<T, bitDepth>(z) << 2) | (impl::separate_bits_4d<T, bitDepth>(w) << 3); }
```

Comment:
imho these should be `template<uint16_t Dims, typename T, uint16_t BitDepth=sizeof(T)*8>`

because that way you only need to write the encode and decode once
```cpp
template<uint16_t Dims, typename T, uint16_t BitDepth=sizeof(T)*8>
vector<T,Dims> decode(const T _morton)
{
    vector<T,Dims> retval;
    for (uint16_t i=0; i<Dims; i++)
       retval[i] = impl::decode<Dims,T,BitDepth>(_morton>>i);
    return retval;
}

template<uint16_t Dims, typename T, uint16_t BitDepth=sizeof(T)*8>
T encode(const vector<T,Dims> coord)
{
    T retval = impl::separate_bits<Dims,T,BitDepth>(coord[0]);
    for (uint16_t i=1; i<Dims; i++)
       retval |= impl::separate_bits<Dims,T,BitDepth>(coord[1])<<i;
    return retval;
}
```